src/ThriftSQL/ImpalaQuery.php

```diff
@@ -19,14 +20,15 @@ public function __construct( $queryStr, $client ) {
   }
 
   public function wait() {
-    $sleeper = new \ThriftSQL\Utils\Sleeper();
+    if ( $this->_ready ) {
+      return $this;
+    }
 
-    // Wait for results
+    // Wait for query to be ready
```

Comment:
`s/query/results/` ?

Reply:
This is part of the problem, sometimes the thrift query handle gets the query state marked as `FINISHED` before the query is ready to return results. Perhaps it gets marked as finished when fragments finish executing but before the gateway aggregates the results?

This can be seen if we run something like `SELECT sleep(5000)` which will return almost immediately even with `wait()` called on it. Then the initial `fetch()` call will take ~5s before it returns results.

Reply:
> I'm not sure if this change is being too clever with trying to wait for the initial fetch as well which might be just an edge case with the UDF.

Yeah my bad, I should have read the whole thing and that there are actually 2 distinct waits ( which your comments explained quite well in fact, I should stop commenting as I go when reviewing and do a first full pass  :D ).

I'd keep it as-is though: should we maybe try it with the cupcakes query that triggered this investigation?